modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy

```diff
@@ -281,7 +281,82 @@ class TaskConfig extends LazyMap implements Cloneable {
 
     int getCpus() {
         final value = get('cpus')
-        value ? value as int : 1  // note: always return at least 1 cpus
+        value ? parseCpusInt(value) : 1  // note: always return at least 1 cpus
+    }
+
+    int getCpuMillis() {
+        final value = get('cpus')
+        value ? parseCpuMillis(value) : 1_000
+    }
+
+    static protected int parseCpusInt(value) {
+        if( value instanceof Integer )
+            return value
+        if( value instanceof Number )
+            return Math.ceil(value.toDouble())
+        if( value !instanceof CharSequence )
+            throw new IllegalArgumentException("Not a valid cpus value: '$value'")
+        // turn millis into an integer unit
+        final str = value.toString()
+        if( str.endsWith('m') ) {
+            try {
+                return Math.ceil((double)str.substring(0,str.length()-1).toInteger() / 1000)
+            }
+            catch (NumberFormatException e) {
+                throw new IllegalArgumentException("Not a valid cpus value: '$value'", e)
+            }
+         // turn float to unit
+        }
+        if( str.contains('.') ) {
+            try {
+                return Math.ceil(str.toFloat())
+            }
+            catch (NumberFormatException e) {
+                throw new IllegalArgumentException("Not a valid cpus value: '$value'", e)
+            }
+        }
+        // it should be an integer value
+        try {
+            return Math.ceil(str.toInteger())
+        }
+        catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Not a valid cpus value: '$value'", e)
+        }
+    }
+
+    static protected int parseCpuMillis(value) {
+        if( value instanceof Integer )
+            return value.toInteger() * 1000
+        if( value instanceof Number )
+            return Math.ceil(value.toDouble() * 1000)
```

Comment:
is `Math.ceil` neeeded in this function if we're just trying to get the millis?

Reply:
Since it's double multipled an integer, the result can be a double, therefore it must be rounded. Then, think it's better to round up to the first integer, to avoid corner cases such as `0.4` => 0

Reply:
oh right, it's rounding up to the next milli